packages/environment-check/src/formatter.ts

```diff
@@ -23,3 +23,36 @@ export function countNumberOfServices(catalogResult?: ODataServiceInfo[]): numbe
 export function getServiceCountText(count: number) {
     return count === 1 ? `${count} service` : `${count} services`;
 }
+
+/**
+ * Returns replacer function that can be used with JSON.stringify to detect
+ * and replace circular structures.
+ *
+ * @example JSON.stringify(object, getCircularReplacer());
+ * @returns - replacer that replaces circular structures
+ */
+export function getCircularReplacer(): (key: string, value: any) => any {
+    const stack: any[] = [];
+    const keys: string[] = [];
+    return function replacer(key, value) {
+        if (stack.length > 0) {
+            const thisPos = stack.indexOf(this);
+            if (thisPos === -1) {
+                stack.push(this);
+                keys.push(key);
+            } else {
+                stack.splice(thisPos + 1);
+                keys.splice(thisPos, Infinity, key);
+            }
+            if (stack.includes(value)) {
+                value =
+                    stack[0] === value
+                        ? '|CIRCULAR STRUCTURE|'
+                        : '|CIRCULAR STRUCTURE .' + keys.slice(0, stack.indexOf(value)).join('.') + '|';
+            }
+        } else {
+            stack.push(value);
+        }
```

Comment:
Use an early return and save one indentation level?

Reply:
Reduces the cognitive complexity too as an upshot

Reply:
changed it to the simplified algorithm